File: python/trees/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        answer = []
        q = deque()
        q.append((root, 0))
        while q:
            current = q.popleft()
            if current[0]:
                if len(answer) <= current[1]:
                    answer.append([])
                answer[current[1]].append(current[0].val)
                q.append((current[0].left, current[1] + 1))
                q.append((current[0].right, current[1] + 1))

        finalanswer = []
        for i in answer:
            finalanswer.append(i[-1])

        return finalanswer

File: c++/trees/binary-tree-right-side-view.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        std::queue<std::pair<TreeNode*, int>> q;
        q.push(std::make_pair(root, 0));
        
        std::vector<int> levels;
        while (q.size() > 0) {
            std::pair<TreeNode*, int> node = q.front();
            q.pop();

            if (!node.first) {
                continue;
            }

            if (node.second == levels.size()) {
                levels.push_back(0);
            }

            levels[node.second] = node.first->val;

            q.push(std::make_pair(node.first->left, node.second + 1));
            q.push(std::make_pair(node.first->right, node.second + 1));
        }

        return levels;
    }
};
